### DISCUSSION TITLE
Determine if files are managed by git-lfs

### DISCUSSION THREAD

**Discussion by ihube** [git-lfs/git-lfs#5313] [General]
I posted this to StackOverflow but with no response so trying here :-).

We have a git repo that is exclusively setup to store large binary files; the intention is that every file will be stored using git-lfs (which is new to most of the team). There are 106564 files.

We are running in to issues where certain team members (graphic artists - who are not git experts) are committing changes without git-lfs installed and we're consequently running in to issues where we see this warning:

Encountered 1 file(s) that should have been pointers, but weren't:

I've written some python code to compare the output of these two commands:

```
git ls-files (list all files)
git lfs ls-files -n (list all files tracked by lfs)
```

The difference between the two lists (once I've dealt with some other issues) in theory provides me with a list of files not tracked by lfs. Before I start going further down the rabbit hole, I'd like to manually confirm on a file basis if my code is working. According to my code I have 1008 not tracked by LFS, if I run the following command on one of the 1008 files I get this output:

`git check-attr -a dirA/dirB/dirC/aFile.ext `

gives:

```
dirA/dirB/dirC/aFile.ext: diff: lfs 
dirA/dirB/dirC/aFile.ext: merge: lfs
dirA/dirB/dirC/aFile.ext: filter: lfs
dirA/dirB/dirC/aFile.ext: crlf: unset
```
In this case the particular file shows up in the “git ls-files” list but not in the “git lfs ls-files” list, implying to me that the file is not tracked by lfs; yet the output of the “git check-attr -a” command leads me to believe otherwise… What am I missing and is there something better than the “git check-attr -a” command to check a file?

I’m using git lfs version as follows on Ubuntu 20.04:

`git-lfs/3.3.0 (GitHub; linux amd64; go 1.19.3)`

I was expecting the following command to tell me this file is not tracked by lfs:

`git check-attr -a dirA/dirB/dirC/aFile.ext`

It doesn't! Ultimately I'm looking to migrate these files to lfs.

**Comment by bk2204** [comment]
Hey,

Let me provide a little context here.  There are three possible states you might have here:

1. Plain Git files (no `filter=lfs`, stored in Git, not as pointer file)
2. Plain LFS files (`filter=lfs`, stored as pointer file)
3. Broken LFS files (`filter=lfs`, stored in Git, not as pointer file)

Both 1 and 2 are fine from a functionality perspective, although you may want some files to be stored in LFS that are not.  State 3 is what leads you to the “should have been pointers but weren't” message, and is not going to work from a functionality perspective.

If your goal is to verify that 3 is not happening, then the easiest way to do that is `git lfs fsck --pointers BASE..HEAD` (as of 3.0.2), which will crawl the commits in the range and fail if any are in state 3.  This is perfect for CI systems (and was specifically created for that reason).

If your goal is that all files are stored in LFS, then you want all files to be in state 2.  In that case, you want the check above, plus verifying that the file is in `git lfs ls-files -n`.  Note that `git lfs ls-files -n` just tells you if the file has `filter=lfs` (that is, it could be in state 2 or state 3), so it doesn't tell you that it's not broken.

Does that explain things a little better?

**Comment by ihube** [comment]
Hi Brian - Thank you for your help; things are much clearer :-).

My goal at the end of the day is to find files in state (3) and fix them; to do this I'm writing some python code as mentioned above that we can run on occasion to fix any files that may have gone bad (state (3)).  So, with the 1008 files I want to individually look at each file figure out if `git lfs fsck --pointers BASE..HEAD` fails and if it does then resolve the issue (I still have to figure out a good way to do that)...

My next question is how to find a reasonable range of BASE and HEAD for each file; HEAD is easy, I guess BASE is a question of deciding how far to go back and then querying at a particular point in time...  I'm still very new to git but have used ClearCase, Perforce, Plastic SCM, SVN over the years!...

Anyway; you've helped a huge amount so once again - Thank You.

**Comment by cov393** [comment]
It seems like you are using the right commands to identify the files that are not tracked by Git LFS. However, there might be some confusion about the output of the ``git check-attr` command.

The output you provided shows that the file `dirA/dirB/dirC/aFile.ext` has the `diff`, `merge`, and `filter` attributes set to `lfs. This means that Git LFS should be used for this file when performing those operations. However, this does not necessarily mean that the file is currently tracked by Git LFS.

To confirm if a file is tracked by Git LFS, you can use the `git lfs ls-files` command. This command should output a list of all files that are currently tracked by Git LFS. If the file `dirA/dirB/dirC/aFile.ext` is not listed in the output of this command, then it is not currently tracked by Git LFS.

To migrate the files to Git LFS, you can use the `git lfs migrate` command. This command will rewrite the Git history to use Git LFS for the specified files. You can specify the files to migrate using various options, such as `--include` and `--exclude` . Be sure to read the documentation for this command carefully and make a backup of your repository before proceeding with the migration.
